sumpy/tools.py

```diff
@@ -123,6 +209,159 @@ def get_closest_cached_mi(self, mi):
+        """
+        See [1] for the algorithm
+
+        [1]: Tausch, J., 2003. The fast multipole method for arbitrary Green's
+             functions. Contemporary Mathematics, 329, pp.307-314.
+        """
+        if not self.is_radial:
+            assert q == 0
+            return MiDerivativeTaker.diff(self, mi)
+
+        try:
+            return self.cache_by_mi_q[(mi, q)]
+        except KeyError:
+            pass
+
+        for i in range(self.dim):
+            if mi[i] == 1:
+                mi_minus_one = list(mi)
+                mi_minus_one[i] = 0
+                mi_minus_one = tuple(mi_minus_one)
+                expr = self.var_list_multiplied[i] * self.diff(mi_minus_one, q=q+1)
+                self.cache_by_mi_q[(mi, q)] = expr
+                return expr
+
+        for i in range(self.dim):
+            if mi[i] >= 2:
+                mi_minus_one = list(mi)
+                mi_minus_one[i] -= 1
+                mi_minus_one = tuple(mi_minus_one)
+                mi_minus_two = list(mi)
+                mi_minus_two[i] -= 2
+                mi_minus_two = tuple(mi_minus_two)
+                expr = (mi[i]-1)*self.diff(mi_minus_two, q=q+1) * self.rscale ** 2
+                expr += self.var_list_multiplied[i] * self.diff(mi_minus_one, q=q+1)
+                expr = add_to_sac(self.sac, expr)
+                self.cache_by_mi_q[(mi, q)] = expr
+                return expr
+
+        assert mi == (0,)*self.dim
+        assert q > 0
+
+        prev_expr = self.diff(mi, q=q-1)
+        # Need to get expr.diff(r)/r, but we can only do expr.diff(x)
+        # Use expr.diff(x) = expr.diff(r) * x / r
+        expr = prev_expr.diff(self.var_list[0])/self.var_list[0]
+        # We need to distribute the division above
+        expr = expr.expand(deep=False)
+        self.cache_by_mi_q[(mi, q)] = expr
+        return expr
+
+
+class HelmholtzDerivativeTaker(RadialDerivativeTaker):
+
+    def diff(self, mi, q=0):
+        import sumpy.symbolic as sym
+        if q < 2 or mi != (0,)*self.dim:
+            return RadialDerivativeTaker.diff(self, mi, q)
+        try:
+            return self.cache_by_mi_q[(mi, q)]
+        except KeyError:
+            pass
+
+        if self.dim == 2:
+            # See https://dlmf.nist.gov/10.6.E6
+            # and https://dlmf.nist.gov/10.6#E1
+            k = self.orig_expr.args[1] / self.r
+            expr = -  2 * (q - 1) * self.diff(mi, q - 1)
+            expr += - k**2 * self.diff(mi, q - 2)
+            expr /= self.r**2
+        else:
+            # See reference [1] in RadialDerivativeTaker.diff
+            k = (self.orig_expr * self.r).args[-1] / sym.I / self.r
+            expr = -(2*q - 1)/self.r**2 * self.diff(mi, q - 1)
+            expr += -k**2 / self.r * self.diff(mi, q - 2)
+        self.cache_by_mi_q[(mi, q)] = expr
+        return expr
+
+
+MiDerivativeTakerWrapper = namedtuple("MiDerivativeTakerWrapper",
+                                      ["taker", "initial_mi"])
```

Comment:
What's the purpose of this? (add a comment maybe)

Reply:
Done in e93c6709f4aa41b3365f6549ad8126505962cdd8